### PR TITLE
Bug 1851999 - Prevent `ping` lifetime metrics on the `events` ping

### PR DESCRIPTION
**Pull request by perrymcmanis144**
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

**Comment by perrymcmanis144** [comment]
hey @badboy here's a draft I'm working on. I figured this one looked like it needed to be solved in the python and not a schema so I did it that way.

just as a point of interest, I tried to run `make install-kotlin-linters` and this appears to require an explicit install of coreutils on mac.

**Comment by badboy** [comment]
> just as a point of interest, I tried to run `make install-kotlin-linters` and this appears to require an explicit install of coreutils on mac.

probably because it uses `sha256sum`? We could change it to `shasum -a 256`

**Comment by badboy** [comment]
Can we implement this as a lint? https://github.com/mozilla/glean_parser/blob/b03df4dbced33d84751135ecb34e9696332228d6/glean_parser/lint.py#L357

**Comment by perrymcmanis144** [comment]
> Can we implement this as a lint?
> 
> https://github.com/mozilla/glean_parser/blob/b03df4dbced33d84751135ecb34e9696332228d6/glean_parser/lint.py#L357

Sure, updated to reflect.

File: glean_parser/lint.py [hunk 1]
```diff
@@ -286,6 +286,22 @@ def check_old_event_api(
...
+    """A non-event metric on the events ping only makes sense if its lifetime
+    is tied to that events ping's lifetime."""
+    if (
+        "events" in metric.send_in_pings
+        and "all_pings" not in metric.send_in_pings
+        and metric.type != metrics.Event
+        and metric.lifetime != metrics.Lifetime.ping
+    ):
+        yield (
+            "Non-event metrics sent on the Events ping should have the ping lifetime"
```
Comment (badboy) [inline glean_parser/lint.py:301]: We usually refer to them by their lowercase name.
```suggestion
            "Non-event metrics sent on the events ping should have the ping lifetime"
```

File: tests/test_lint.py [hunk 2]
```diff
@@ -279,6 +279,27 @@ def test_expires_too_far_in_the_future():
     assert set(["EXPIRATION_DATE_TOO_FAR"]) == set(v.check_name for v in nits)
 
 
+def test_invalid_lifetime_for_metric_on_events_ping():
+    """Test that a `expires` dates too far in the future generates warnings"""
```
Comment (badboy) [inline tests/test_lint.py:283]: Copy-pasta. Needs an update.

File: glean_parser/lint.py [hunk 3]
```diff
@@ -286,6 +286,22 @@ def check_old_event_api(
...
+def check_metric_on_events_lifetime(
+    metric: metrics.Metric, parser_config: Dict[str, Any]
+) -> LintGenerator:
+    """A non-event metric on the events ping only makes sense if its lifetime
+    is tied to that events ping's lifetime."""
+    if (
+        "events" in metric.send_in_pings
+        and "all_pings" not in metric.send_in_pings
+        and metric.type != metrics.Event
+        and metric.lifetime != metrics.Lifetime.ping
```
Comment (badboy) [inline glean_parser/lint.py:298]: Wait, we want to _exclude_ ping-lifetime. `application` and user` lifetime are ok, these are static over time.

File: glean_parser/lint.py (hunk 1 above)
Comment (badboy) [reply glean_parser/lint.py:301]: As said above, `ping` lifetime should be disallowed, `application` is what should be prefered.

File: tests/test_lint.py [hunk 4]
```diff
@@ -279,6 +279,27 @@ def test_expires_too_far_in_the_future():
...
+    contents = [
+        {
+            "user_data": {
+                "invalid_lifetime": {
+                    "type": "counter",
+                    "lifetime": "application",
```
Comment (badboy) [inline tests/test_lint.py:289]: As said this should pass, `ping` should make it fail

**Review by badboy — CHANGES_REQUESTED** [review]

File: glean_parser/lint.py [hunk 5]
```diff
@@ -286,6 +286,23 @@ def check_old_event_api(
...
 
+def check_metric_on_events_lifetime(
+    metric: metrics.Metric, parser_config: Dict[str, Any]
+) -> LintGenerator:
+    """A non-event metric on the Events ping only makes sense if its value
+    is immutable over the life of the ping."""
+    if (
+        "events" in metric.send_in_pings
+        and "all_pings" not in metric.send_in_pings
+        and metric.type != metrics.Event
```
Comment (badboy) [inline glean_parser/lint.py:297]: `metric.type` is a string, we can't compare that with a class.
```suggestion
        and metric.type != "event"
```

or (I think?)

```suggestion
        and not isinstance(metric, metrics.Event)
```

the former is more stringy, but IMO reads better

**Review by badboy — CHANGES_REQUESTED** [review]
See inline comments for one.
There are more errors after that because in our test data we seem to have a couple of metrics that trigger that linter and thus now fail. The proper way is to make them use either another metric type in the test or set the correct lifetime

**Review by badboy — APPROVED** [review]

**Pull request merged**